File: systems-for-design-and-implementation/labs/lab3/LibraryServices/src/library/services/ILibraryServer.scala
package library.services

import java.util

import library.model.{Book, BookLoan, User}

trait ILibraryServer
{
    @throws[LibraryException]
    def login(user: User, client: ILibraryClient): User

    @throws[LibraryException]
    def logout(user: User, client: ILibraryClient)

    @throws[LibraryException]
    def getAvailableBooks: util.List[Book]

    @throws[LibraryException]
    def getBorrowedBy(user: User): util.List[Book]

    @throws[LibraryException]
    def searchByTitle(title: String): util.List[Book]

    @throws[LibraryException]
    def getAllLoans: util.List[BookLoan]

    @throws[LibraryException]
    def borrowBook(book: Book, user: User)

    @throws[LibraryException]
    def returnBook(book_id: Int, user_id: Int)
}

File: systems-for-design-and-implementation/labs/lab2/LibraryServices/src/library/services/ILibraryServer.java
package library.services;

import library.model.Book;
import library.model.BookLoan;
import library.model.User;

import java.util.List;

public interface ILibraryServer
{
    User login(User user, ILibraryClient client) throws LibraryException;

    void logout(User user, ILibraryClient client) throws LibraryException;

    List<Book> getAvailableBooks() throws LibraryException;

    List<Book> getBorrowedBy(User user) throws LibraryException;

    List<Book> searchByTitle(String title) throws LibraryException;

    List<BookLoan> getAllLoans() throws LibraryException;

    void borrowBook(Book book, User user) throws LibraryException;

    void returnBook(int book_id, int user_id) throws LibraryException;
}
